packages/hardhat-core/src/internal/hardhat-network/provider/vm/dual.ts

```diff
@@ -89,42 +100,102 @@ export class DualModeAdapter implements VMAdapter {
   }
 
   public async getStateRoot(): Promise<Buffer> {
-    return this._ethereumJSAdapter.getStateRoot();
+    const ethereumJSRoot = await this._ethereumJSAdapter.getStateRoot();
+    const rethnetRoot = await this._rethnetAdapter.getStateRoot();
+
+    if (!ethereumJSRoot.equals(rethnetRoot)) {
+      console.trace(
+        `Different state root: ${ethereumJSRoot.toString(
+          "hex"
+        )} !== ${rethnetRoot.toString("hex")}`
+      );
+      throw new Error("Different state root");
+    }
+
+    return rethnetRoot;
   }
 
   public async getAccount(address: Address): Promise<Account> {
-    return this._ethereumJSAdapter.getAccount(address);
+    const ethereumJSAccount = await this._ethereumJSAdapter.getAccount(address);
+    const rethnetAccount = await this._rethnetAdapter.getAccount(address);
+
+    assertEqualAccounts(address, ethereumJSAccount, rethnetAccount);
+
+    return ethereumJSAccount;
   }
 
   public async getContractStorage(
     address: Address,
     key: Buffer
   ): Promise<Buffer> {
-    return this._ethereumJSAdapter.getContractStorage(address, key);
+    const ethereumJSStorageSlot =
+      await this._ethereumJSAdapter.getContractStorage(address, key);
+
+    const rethnetStorageSlot = await this._rethnetAdapter.getContractStorage(
+      address,
+      key
+    );
+
+    if (
+      !(
+        ethereumJSStorageSlot.length === 0 &&
+        rethnetStorageSlot.equals(Buffer.from([0x00]))
+      ) &&
+      !ethereumJSStorageSlot.equals(rethnetStorageSlot)
+    ) {
```

Comment:
I found this very hard to understand. What about doing something like:

```js
if (!ethereumJSStorageSlot.equals(rethnetStorageSlot)) {
  // we only throw if any of the returned values was non-empty
  if (ethereumJSStorageSlot.length !== 0 || !rethnetStorageSlot.equals(Buffer.from([0x00]))) {
    // throw here
  }
}
```

I'm not sure if the comment is 100% accurate, but I guess that's the idea?

Reply:
Yeah, that makes sense. As far as I understand the spec, you're supposed to return zero if a storage slot is empty.

Ethereumjs just returns an empty buffer in that case, hence why the two implementations diverge and we need to check this. I'm not sure if that helps you write a better comment